### PR TITLE
[PURR][#1343]Implementation of ORCID Create-on-Demand process on Hubzero

### PR DESCRIPTION
**Pull request by kuang5**
The changes are about implementation of the ORCID create-on-demand process on Hubzero. Please find the details about requirements and implementation in the document attached to PURR ticket #1343.
Thank you, Jerry

### CODE REVIEW

File: core/components/com_members/models/member.php [hunk 1]
```diff
@@ -542,4 +544,103 @@ public static function clearTerms()
...
 		return $query->execute();
 	}
+	
+	/**
+	* check if user has registered with system or not
+	*
+	* @param   string   $name
+	* @return  boolean
+	*/
+	public static function profileExists($name)
```
Comment (kevinwojo) [inline core/components/com_members/models/member.php]: The `User::getInstance()` method returns a User object that achieves the same effect.

File: core/components/com_members/models/member.php [hunk 2]
```diff
@@ -542,4 +544,103 @@ public static function clearTerms()
...
+	}
+	
+	/**
+	* Save ORCID ID to file orcidRecords
+	*
+	* @param   string   $name
+	* @param   string   $orcid
+	* @return  void
+	*/
+	public static function saveOrcidToFile($name, $orcid)
```
Comment (kevinwojo) [inline core/components/com_members/models/member.php]: Why is this needed?

File: core/components/com_members/models/member.php [hunk 3]
```diff
@@ -542,4 +544,103 @@ public static function clearTerms()
...
+		}
+	}
+	
+	/**
+	* Get orcid from orcidRecords
+	*
+	* @param   string   $name
+	* @return  orcid ID
+	*/
+	public static function getORCIDFromFile($name)
```
Comment (kevinwojo) [inline core/components/com_members/models/member.php]: This should be able to be accomplished without writing anything to the filesystem. Session data perhaps? Cookie?

File: core/components/com_members/site/controllers/register.php [hunk 4]
```diff
@@ -555,7 +555,7 @@ public function updateTask()
 	 * @return  void
 	 */
 	public function createTask()
-	{
+	{
```
Comment (kevinwojo) [inline core/components/com_members/site/controllers/register.php]: trailing space

File: core/components/com_members/site/controllers/register.php [hunk 5]
```diff
@@ -1513,5 +1520,5 @@ protected function _buildTitle()
 			$title = Lang::txt('COM_MEMBERS_REGISTER');
 		}
 		\Document::setTitle($title);
-	}
+	}
```
Comment (kevinwojo) [inline core/components/com_members/site/controllers/register.php]: trailing space

File: core/components/com_members/site/views/orcid/tmpl/display.php [hunk 6]
```diff
@@ -143,10 +146,13 @@
 					<h4><?php echo Lang::txt('Create a new ORCID'); ?></h4>
 					<div class="grid nobreak">
 						<div class="col span8">
-							<p><?php echo Lang::txt('If you can\'t find your ID or would like to create one, click the "Create new ORCID" button.'); ?></p>
+							<p><?php echo Lang::txt('If you can\'t find your ID or would like to create one, click the "Create or Connect your ORCID iD" button.'); ?></p>
```
Comment (kevinwojo) [inline core/components/com_members/site/views/orcid/tmpl/display.php]: Language string preferred: https://hubzero.org/documentation/2.1.0/webdevs/components.languages

**Comment by kevinwojo** [comment]
The following style errors have been reported. We are incrementally increasing the readability of our code, so any file that you make changes to is ran against the PHP CodeSniffer rules defined in https://github.com/hubzero/standards. 

Documentation on how to setup your development environment to check your own code can be found here: https://hubzero.org/documentation/2.1.0/webdevs/index.contributions#externalcontributionprocess

```FILE: ...ubzero/hubzero-cms/core/components/com_members/models/member.php
----------------------------------------------------------------------
FOUND 48 ERRORS AFFECTING 39 LINES
----------------------------------------------------------------------
  43 | ERROR | [ ] Trailing white space is not allowed.
 547 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
 547 | ERROR | [ ] Trailing white space is not allowed.
 549 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 550 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 551 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 552 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 553 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 570 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
 570 | ERROR | [ ] Trailing white space is not allowed.
 572 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 573 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 574 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 575 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 576 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 577 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 585 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
 585 | ERROR | [ ] Trailing white space is not allowed.
 587 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 588 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 589 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 590 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 591 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 592 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 594 | ERROR | [ ] Trailing white space is not allowed.
 599 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
 599 | ERROR | [ ] Trailing white space is not allowed.
 601 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
 601 | ERROR | [ ] Trailing white space is not allowed.
 603 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
 603 | ERROR | [ ] Trailing white space is not allowed.
 606 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
 606 | ERROR | [ ] Trailing white space is not allowed.
 608 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 609 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 610 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 611 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 612 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 614 | ERROR | [ ] Trailing white space is not allowed.
 619 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
 619 | ERROR | [ ] Trailing white space is not allowed.
 632 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
 632 | ERROR | [ ] Trailing white space is not allowed.
 634 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 635 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 636 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 637 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
 638 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
----------------------------------------------------------------------
PHPCBF CAN FIX THE 27 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: ...ro-cms/core/components/com_members/site/controllers/redirect.php
----------------------------------------------------------------------
FOUND 29 ERRORS AFFECTING 19 LINES
----------------------------------------------------------------------
  43 | ERROR | [x] Opening class brace must be on a line by itself
  43 | ERROR | [ ] Trailing white space is not allowed.
  45 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
  46 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
  47 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
  48 | ERROR | [ ] Tabs must be used to indent lines; line cannot
     |       |     begin with spaces.
  49 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
  49 | ERROR | [ ] Trailing white space is not allowed.
  51 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
  52 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
  53 | ERROR | [x] Expected 2 space(s) before asterisk; 1 found
  54 | ERROR | [ ] Tabs must be used to indent lines; line cannot
     |       |     begin with spaces.
  55 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
  55 | ERROR | [ ] Trailing white space is not allowed.
  65 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
  65 | ERROR | [ ] Trailing white space is not allowed.
  80 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
  80 | ERROR | [ ] Trailing white space is not allowed.
  85 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
  85 | ERROR | [ ] Trailing white space is not allowed.
  94 | ERROR | [ ] Trailing white space is not allowed.
 106 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
 106 | ERROR | [ ] Trailing white space is not allowed.
 118 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
 118 | ERROR | [ ] Trailing white space is not allowed.
 129 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
 129 | ERROR | [ ] Trailing white space is not allowed.
 143 | ERROR | [ ] Blank lines containing indentation are not
     |       |     allowed.
 143 | ERROR | [ ] Trailing white space is not allowed.
----------------------------------------------------------------------
PHPCBF CAN FIX THE 7 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: ...ro-cms/core/components/com_members/site/controllers/register.php
----------------------------------------------------------------------
FOUND 6 ERRORS AFFECTING 4 LINES
----------------------------------------------------------------------
  558 | ERROR | Trailing white space is not allowed.
  791 | ERROR | Blank lines containing indentation are not allowed.
  791 | ERROR | Trailing white space is not allowed.
  798 | ERROR | Blank lines containing indentation are not allowed.
  798 | ERROR | Trailing white space is not allowed.
 1523 | ERROR | Trailing white space is not allowed.
----------------------------------------------------------------------```

File: core/components/com_members/models/member.php (hunk 2 above)
Comment (kuang5) [inline core/components/com_members/models/member.php]: My design is that when there isn't such user's profile record in database, save the generated ORCID ID in temporary file. When user creates account, which means corresponding record creates in database at the same time, it reads ORCID from the temporary file and save it in the database.

File: core/components/com_members/models/member.php (hunk 3 above)
Comment (kuang5) [inline core/components/com_members/models/member.php]: If using session data or cookies, but user closes the registration page or user profile page, do the data still exist in session or cookie?

Comment (kuang5) [inline core/components/com_members/models/member.php]: If user happens to clear the browser cookie, can we still use the stored data?

File: core/components/com_members/models/member.php (hunk 2 above)
Comment (zooley) [inline core/components/com_members/models/member.php]: The `#__xprofiles` table should not be used. It currently only exists for legacy reasons. All profile data is stored in `#__user_profiles`.

File: core/components/com_members/site/controllers/redirect.php [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+			}
+			else
+			{
+				Member::saveOrcidToFile($this->_userName, $this->_userOrcidID);
+			}
+		}
+		
+		$this->view->display();
+	}
+}
```
Comment (zooley) [inline core/components/com_members/site/controllers/redirect.php]: This entire controller only handles ORCID related data. Why weren't the methods just added to the existing ORCID controller?

File: core/components/com_members/site/views/orcid/tmpl/display.php [hunk 8]
```diff
@@ -143,10 +146,13 @@
...
 						</div>
 						<div class="col span4 omega">
-							<p><a id="create-orcid" class="btn" href="https://orcid.org/register" target="_blank"><?php echo Lang::txt('Create new ORCID'); ?></a></p>
+							<p><a id="create-orcid" class="btn" href="https://<?php if ($this->config->get('orcid_service', 'members') == 'sandbox') { echo Lang::txt('sandbox.'); }?>orcid.org/oauth/authorize?client_id=<?php echo $clientID ?>
+							&response_type=code&scope=/authenticate&redirect_uri=<?php echo urlencode($redirectURI) ?>&family_names=<?php echo $this->escape($lname);?>
```
Comment (zooley) [inline core/components/com_members/site/views/orcid/tmpl/display.php]: `echo Lang::txt('sandbox.');` - Remove `Lang::txt()`. It's not needed and could potentially cause issues.

Ampersands (&) in the URL need to be encoded `&amp;`.

File: core/components/com_members/models/member.php (hunk 2 above)
Comment (kuang5) [inline core/components/com_members/models/member.php]: Thank you, Shawn. Will update and resubmit the patch along with other fixes.

File: core/components/com_members/site/controllers/redirect.php (hunk 7 above)
Comment (kuang5) [inline core/components/com_members/site/controllers/redirect.php]: This controller has a corresponding landing page view (display.php) to show user's ORCID ID after it is created or some indication message if user denies the authorization. If I add the methods to ORCID controller, I don't know whether it is possible to manage this view with the existing ORCID view. Is it possible to put the two views together in one display.php and show the landing page view according to whether authorization code is posted to Hubzero or not?

Comment (zooley) [inline core/components/com_members/site/controllers/redirect.php]: A controller can have any number of views with different names. Typically, the name of the view corresponds to the task name. For example, you add these methods to the ORCID controller, rename this `displayTask` to `redirectTask`, rename its view to `redirect.php` and add the view to the orchid views.

Comment (kuang5) [inline core/components/com_members/site/controllers/redirect.php]: Thank you, Shawn. Will update the code.

**Comment by kuang5** [comment]
To whom it may concern,

I see that there are style errors appears for FILE: ..cms/core/components/com_resources/site/controllers/resources.php

However, I don't make any changes on this file. Is there any way to remove this error so the request can be successfully submitted? Or do I have to fix the style errors in the resources.php?

**Comment by kuang5** [comment]
The style errors infile: ..cms/core/components/com_resources/site/controllers/resources.php have been fixed.

File: core/components/com_members/site/controllers/orcid.php [hunk 9]
```diff
@@ -45,6 +45,28 @@
...
 {
 	/**
+	* user's name
+	*
+	*/
+	protected $_userName;
+	
+	/**
+	* user's ORCID ID
+	*
```
Comment (zooley) [inline core/components/com_members/site/controllers/orcid.php]: Include variable type (e.g., "@var  string")

File: core/components/com_members/site/controllers/orcid.php [hunk 10]
```diff
@@ -45,6 +45,28 @@
 class Orcid extends SiteController
 {
 	/**
+	* user's name
+	*
```
Comment (zooley) [inline core/components/com_members/site/controllers/orcid.php]: Include variable type (e.g., "@var  string")

File: core/components/com_members/models/member.php [hunk 11]
```diff
@@ -1,545 +1,574 @@
...
+	* @return  int user ID
+	*/
+	public static function getUserID($name)
+	{
+		$db = App::get('db');
+		$sql = 'SELECT id FROM #__users' . ' WHERE name = ' . '"' . $name . '"';
+		$db->setQuery($sql);
+		$id = (int)$db->loadResult();
+		return $id;
+	}
```
Comment (zooley) [inline core/components/com_members/models/member.php]: Incorrect way to do this. Use the ORM. Do not use raw SQL.
```
return self::all()
    ->whereEquals('name', $name)
    ->row()
    ->get('id');
```

Comment (zooley) [inline core/components/com_members/models/member.php]: This method has no use inside of the `create` task of the registration controller. That is, it's used to check if an account already exists for a name inside a method that is explicitly for creating new accounts. You have to already assume an account does not exist!

Also, it could very easily fail as it doesn't check inclusion of middle name or not, abbreviations (e.g., "J. Smith"), or the fact that there can be multiple people with the same names.

I don't quite see the point of this method. If the user is logged in, you use `User::get('id')`. If the user is not logged in, then they are registering and they have no ID.

File: core/components/com_members/models/member.php [hunk 12]
```diff
@@ -1,545 +1,574 @@
...
+	
+	/**
+	* Save ORCID ID to Session
+	*
+	* @param   string   $name
+	* @param   string   $orcid
+	* @return  void
+	*/
+	public static function saveOrcidToSession($name, $orcid)
```
Comment (zooley) [inline core/components/com_members/models/member.php]: Remove this from the `Member` model. It is specific to the orcid process and controller (where it should be).

File: core/components/com_members/models/member.php [hunk 13]
```diff
@@ -1,545 +1,574 @@
...
+		return $orcid;
+	}
+	
+	/**
+	* Save orcid to #_user_profiles
+	*
+	* @param   string   $name, $orcid
+	* @return  void
+	*/
+	public static function saveORCIDToProfile($name, $orcid)
```
Comment (zooley) [inline core/components/com_members/models/member.php]: Remove this from the Member model. It is specific to the orcid process and controller (where it should be).

File: core/components/com_members/models/member.php [hunk 14]
```diff
@@ -1,545 +1,574 @@
...
-		$tbl = self::blank();
-
-		$query = $tbl->getQuery()
-			->update($tbl->getTableName())
-			->set(array('usageAgreement' => 0));
-
-		return $query->execute();
-	}
-}
+<?php
```
Comment (zooley) [inline core/components/com_members/models/member.php]: This entire file shows far too many changes. Changes to whitespace, etc. that shouldn't be there.

File: core/components/com_members/models/member.php [hunk 15]
```diff
@@ -1,545 +1,574 @@
...
+		Session::set($sesID, $orcid);
+	}
+	
+	/**
+	* Get orcid from Session
+	*
+	* @param   string   $name
+	* @return  orcid ID
+	*/
+	public static function getORCIDFromSession($name)
```
Comment (zooley) [inline core/components/com_members/models/member.php]: There's no reason for this to be a method; It's only used in one spot. Remove this from the Member model.

File: core/components/com_members/models/member.php [hunk 16]
```diff
@@ -1,545 +1,574 @@
...
+		return $id;
+	}
+	
+	/**
+	* check if user has registered with system or not
+	*
+	* @param   string   $name
+	* @return  boolean
+	*/
+	public static function userExists($name)
```
Comment (zooley) [inline core/components/com_members/models/member.php]: Unnecessary abstraction. Remove this from the Member model.
```
if (!Member::getUserID($name))
{
    // do stuff
}
```

File: core/components/com_members/models/member.php (hunk 11 above)
Comment (kuang5) [inline core/components/com_members/models/member.php]: Hi Shawn, thank you for pointing out the use issue. 
As we know that there are two scenarios that user may create ORCID. One is creating ORCID through the ORCID user interface on the account profile page. Another is creating ORCID through the ORCID user interface on account register page. For the first scenario, when ORCID is created it can be directly saved in database since such user exists in system. For the second one, the ORCID ID has to be saved in temporary location after it is created since there isn't such user in system. When user creates the account by clicking the "account create" button, we read the orcid from temporary location and save it in database.
So the reason that checking whether there is an account existing for a name in the account registering process is that we want to get the user's ID in order to save the orcid id as a record in database. (User creates ORCID ID through the ORCID user interface when registering an account). Actually here I think we don't have to check because it is in such user's account creation process, and at that exact time, the account id is already created in system. We just have to read the orcid from session and save it along with user's id as a record in database. 
The reason that middle name is not taken into account is that the ORCID registration only needs first name and last name (see their user interface on: https://orcid.org/register). Also, according to ORCID workflow, when ORCID is created, the information we only know are first name, last name, and ORCID ID. So we have to user the name to get the user's id, and then create a orcid record for such user in database. I agree with you that there is failure risk if middle name is not included for account check. 
Other functionalities like saveOrcidToSession, saveORCIDToProfile, getORCIDFromSession, userExists are defined and used to complete the requirement for saving the ORCID in database.

Comment (kuang5) [inline core/components/com_members/models/member.php]: I have asked ORCID support if they can add the user's email address in the ORCID creation response. If it can be true, then we can check if such user exists in system or not by the email address.

Comment (zooley) [inline core/components/com_members/models/member.php]: I'm well aware of how the scenarios of ORCID registration work.

ORCID may only use first and last name but the hub registration uses more than that and you're looking up a *hub* account by their name. The name stored on the hub will _include middle name_. So, attempting to load an account `FROM jos_users WHERE name='$xregistration->get('givenName', '') . ' ' . $xregistration->get('surname', '')'`will not always work.

_It is also entirely unnecessary._ As stated before, when the user is logged-in (and looking up their ORCID) you already have access to their ID through `User::get('id')`. No lookup is necessary. When the user is logged out, the data should be saved to and retrieved from the user's session. Again, no lookup by name is needed there. The session is always tied to the current user, logged in or not. Simply using `Session::set('orcid', $ocid)` and `Session::get('orcid')` should be sufficient.

File: core/components/com_members/models/member.php (hunk 12 above)
Comment (kuang5) [inline core/components/com_members/models/member.php]: OK, will just use session API instead of the functionality.

File: core/components/com_members/models/member.php (hunk 15 above)
Comment (kuang5) [inline core/components/com_members/models/member.php]: OK, this function will be removed.

File: core/components/com_members/models/member.php (hunk 16 above)
Comment (kuang5) [inline core/components/com_members/models/member.php]: I might be wrong but please feel free to let me know. In order to build user's orcid record (id, user_id, profile_key, profile_value, ordering, access), we have to get the user id. It can be achieved if user is logged in or in last step of the account creation process. I found that $user->get('id') can return the user's id after user's record is created in the account creation process.

Comment (kuang5) [inline core/components/com_members/models/member.php]: The user exist check will be removed.

File: core/components/com_members/models/member.php (hunk 13 above)
Comment (kuang5) [inline core/components/com_members/models/member.php]: The saveORCIDToProfile is to implement the requirement for saving the ORCID ID in jos_user_profiles. So it is necessary to retain it.

Comment (zooley) [inline core/components/com_members/models/member.php]: I repeat, it is specific to the orcid process and controller (_where it should be_). Move it out of the `Member` model.

Comment (kuang5) [inline core/components/com_members/models/member.php]: Ok, is it ok to move this function to ORCID controller? But It may not fit there because controller doesn't interact with database. This function is invoked only for the scenario. when user is logged in and an orcid is created. Please advise how I should implement "saving orcid in database" requirement if all of above doesn't meet the coding requirement.

**Comment by kuang5** [comment]
Hi Shawn and Kevin,

I have modified the code according to the review result and local test shows they work fine. The changes have been committed. Please review them.

Thank you,
Jerry

File: core/components/com_members/site/views/orcid/tmpl/redirect.php [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+// No direct access
+defined('_HZEXEC_') or die();
+
+?>
+<section class="main section">
+      <div class="grid nobreak">
+		<?php if (Request::getVar('code')) { ?>
+		<h1><?php echo Lang::txt('COM_MEMBERS_REDIRECT_ORCID_THANK_YOU'); ?><?php echo $this->userName; ?><?php echo Lang::txt('COM_MEMBERS_REDIRECT_ORCID_EXCLAMATION_MARK'); ?></h1>
+		<br>
+		<p><?php echo Lang::txt('COM_MEMBERS_REDIRECT_ORCID_YOUR_ORCID'); ?><img src="<?php echo Request::root()?>/core/components/com_members/site/assets/img/orcid_16x16.png" class="logo" width='16' height='16' alt="iD"/> <?php echo Lang::txt('COM_MEMBERS_REDIRECT_ORCID_IS'); ?> <?php echo $this->userORCID; ?></p>
```
Comment (zooley) [inline core/components/com_members/site/views/orcid/tmpl/redirect.php]: Do not use single quotes on element attributes.
```
<img ... width="16" height="16"
```

File: core/components/com_members/models/member.php (hunk 14 above)
Comment (zooley) [inline core/components/com_members/models/member.php]: Diff still shows almost this entire file as having changed. Different (i.e., non-Unix) line endings?

File: core/components/com_members/site/views/orcid/tmpl/redirect.php (hunk 17 above)
Comment (kuang5) [inline core/components/com_members/site/views/orcid/tmpl/redirect.php]: This one is fixed.

File: core/components/com_members/models/member.php (hunk 14 above)
Comment (kuang5) [inline core/components/com_members/models/member.php]: This one is fixed.

**Comment by kuang5** [comment]
Hi Shawn and Kevin,

I fixed the styling error in member.php and other coding convention issues. Please let me know if there is still any issues.

Jerry

**Comment by kuang5** [comment]
Hi Shawn and Kevin,

Could you please tell me why this request is still open? If there are any changes I need to make, please let me know.

Thank you,
Jerry

**Comment by kevinwojo** [comment]
Due to the age and the quality of this merge request, I am closing this. We can incorporate this code into the specific hub (PURR) by putting this in the app directory.
